src/Feldspar/Compiler/Backend/C/CodeGeneration.hs

```diff
@@ -75,14 +75,12 @@ instance CodeGen (Entity ())
       | otherwise = start <> semi
         where
          start
-           | loopBody  = text ("LOOP_BODY_" ++ show (length inParams + length outs - 1))
+           | loopBody  = text ("LOOP_BODY_" ++ show (length inParams - 1))
```

Comment:
Doesn't this change the arity of Wool loop bodies?

Reply:
No, because the variables that would have been in outParams are now included in inParams. It might be a good idea to rename inPrams, but I judged it best to leave that to a separate refactoring commit.

Reply:
I don't like the return value handling we currently have but the distinction between inParams and outParams means we separate the variables that are updated from those that aren't. I realize we can get that information back through an analysis but that was where we started and that feldspar-compiler was both larger and more fragile.

Reply:
Well, we do not currently use that information, and it is somewhat dubious anyway. For instance, we update input arrays using initArray, so they would be an exception. And in addition, the information gets lost in round-tripping. A more natural way to handle this is to use returns.

Reply:
Should have been "fast returns".

Reply:
Well, initArray and friends are similar to memcpy in the sense that the compiler has built-in knowledge about them. The reasonable interface to something that allocates memory for an array is to return a pointer to the memory but array handling is opaque in Program and we didn't have pointers in early versions of Program. I'm hoping this will be improved with the array representation changes.

I find the round-tripping argument pretty convincing, either we fix that by adding annotations to parameters so things can round-trip or we remove the "feature" and use regular returns. Using a regular return is a minor change and gives us the expected semantics which seems like a better idea. I'll merge the gold files first so we can get this change tested with the added tests.

Reply:
I'm all for regular returns. So you mean you want a patch making regular returns the default, then this one?

Reply:
Sorry, I meant a separate patch that adds a gold test for OpenMP and another gold test for Wool so we know the generated code is the same before/after this patch.

You might be right about making regular returns the default though, do you see any reason for not doing that?

Reply:
Testing that the no-way-out patch does not corrupt Wool and OpenMP programs seems sensible. Will do. Let's just hope we do not discover that they were already broken...

I do not think there are any particular problems with fast returns. The major difference I can see is that the marshaling code will free stuff allocated by the Feldspar code, but it does this already for arrays (since their memory is typically allocated by initArray()).